// @azure/storage-blob/dist/commonjs/generated/src/storageClient.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var storageClient_exports = {};
__export(storageClient_exports, {
  StorageClient: () => StorageClient
});
module.exports = __toCommonJS(storageClient_exports);
var coreHttpCompat = __toESM(require("@azure/core-http-compat"));
var import_operations = require("./operations/index.js");
class StorageClient extends coreHttpCompat.ExtendedServiceClient {
  url;
  version;
  /**
   * Initializes a new instance of the StorageClient class.
   * @param url The URL of the service account, container, or blob that is the target of the desired
   *            operation.
   * @param options The parameter options
   */
  constructor(url, options) {
    if (url === void 0) {
      throw new Error("'url' cannot be null");
    }
    if (!options) {
      options = {};
    }
    const defaults = {
      requestContentType: "application/json; charset=utf-8"
    };
    const packageDetails = `azsdk-js-azure-storage-blob/12.32.0`;
    const userAgentPrefix = options.userAgentOptions && options.userAgentOptions.userAgentPrefix ? `${options.userAgentOptions.userAgentPrefix} ${packageDetails}` : `${packageDetails}`;
    const optionsWithDefaults = {
      ...defaults,
      ...options,
      userAgentOptions: {
        userAgentPrefix
      },
      endpoint: options.endpoint ?? options.baseUri ?? "{url}"
    };
    super(optionsWithDefaults);
    this.url = url;
    this.version = options.version || "2026-04-06";
    this.service = new import_operations.ServiceImpl(this);
    this.container = new import_operations.ContainerImpl(this);
    this.blob = new import_operations.BlobImpl(this);
    this.pageBlob = new import_operations.PageBlobImpl(this);
    this.appendBlob = new import_operations.AppendBlobImpl(this);
    this.blockBlob = new import_operations.BlockBlobImpl(this);
  }
  service;
  container;
  blob;
  pageBlob;
  appendBlob;
  blockBlob;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StorageClient
});
