// @azure/storage-blob/dist/commonjs/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  AnonymousCredential: () => import_storage_common.AnonymousCredential,
  AnonymousCredentialPolicy: () => import_storage_common.AnonymousCredentialPolicy,
  BaseRequestPolicy: () => import_storage_common.BaseRequestPolicy,
  BlockBlobTier: () => import_models.BlockBlobTier,
  Credential: () => import_storage_common.Credential,
  CredentialPolicy: () => import_storage_common.CredentialPolicy,
  Pipeline: () => import_Pipeline.Pipeline,
  PremiumPageBlobTier: () => import_models.PremiumPageBlobTier,
  RestError: () => import_core_rest_pipeline.RestError,
  StorageBlobAudience: () => import_models.StorageBlobAudience,
  StorageBrowserPolicy: () => import_storage_common.StorageBrowserPolicy,
  StorageBrowserPolicyFactory: () => import_storage_common.StorageBrowserPolicyFactory,
  StorageOAuthScopes: () => import_Pipeline.StorageOAuthScopes,
  StorageRetryPolicy: () => import_storage_common.StorageRetryPolicy,
  StorageRetryPolicyFactory: () => import_storage_common.StorageRetryPolicyFactory,
  StorageRetryPolicyType: () => import_storage_common.StorageRetryPolicyType,
  StorageSharedKeyCredential: () => import_storage_common.StorageSharedKeyCredential,
  StorageSharedKeyCredentialPolicy: () => import_storage_common.StorageSharedKeyCredentialPolicy,
  generateAccountSASQueryParameters: () => import_AccountSASSignatureValues.generateAccountSASQueryParameters,
  generateBlobSASQueryParameters: () => import_BlobSASSignatureValues.generateBlobSASQueryParameters,
  getBlobServiceAccountAudience: () => import_models.getBlobServiceAccountAudience,
  isPipelineLike: () => import_Pipeline.isPipelineLike,
  logger: () => import_log.logger,
  newPipeline: () => import_Pipeline.newPipeline
});
module.exports = __toCommonJS(src_exports);
var import_core_rest_pipeline = require("@azure/core-rest-pipeline");
__reExport(src_exports, require("./BlobServiceClient.js"), module.exports);
__reExport(src_exports, require("./Clients.js"), module.exports);
__reExport(src_exports, require("./ContainerClient.js"), module.exports);
__reExport(src_exports, require("./BlobLeaseClient.js"), module.exports);
__reExport(src_exports, require("./sas/AccountSASPermissions.js"), module.exports);
__reExport(src_exports, require("./sas/AccountSASResourceTypes.js"), module.exports);
__reExport(src_exports, require("./sas/AccountSASServices.js"), module.exports);
var import_AccountSASSignatureValues = require("./sas/AccountSASSignatureValues.js");
__reExport(src_exports, require("./BlobBatch.js"), module.exports);
__reExport(src_exports, require("./BlobBatchClient.js"), module.exports);
__reExport(src_exports, require("./sas/BlobSASPermissions.js"), module.exports);
var import_BlobSASSignatureValues = require("./sas/BlobSASSignatureValues.js");
__reExport(src_exports, require("./sas/ContainerSASPermissions.js"), module.exports);
var import_models = require("./models.js");
var import_Pipeline = require("./Pipeline.js");
var import_storage_common = require("@azure/storage-common");
__reExport(src_exports, require("./sas/SASQueryParameters.js"), module.exports);
__reExport(src_exports, require("./generatedModels.js"), module.exports);
var import_log = require("./log.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AnonymousCredential,
  AnonymousCredentialPolicy,
  BaseRequestPolicy,
  BlockBlobTier,
  Credential,
  CredentialPolicy,
  Pipeline,
  PremiumPageBlobTier,
  RestError,
  StorageBlobAudience,
  StorageBrowserPolicy,
  StorageBrowserPolicyFactory,
  StorageOAuthScopes,
  StorageRetryPolicy,
  StorageRetryPolicyFactory,
  StorageRetryPolicyType,
  StorageSharedKeyCredential,
  StorageSharedKeyCredentialPolicy,
  generateAccountSASQueryParameters,
  generateBlobSASQueryParameters,
  getBlobServiceAccountAudience,
  isPipelineLike,
  logger,
  newPipeline,
  ...require("./BlobServiceClient.js"),
  ...require("./Clients.js"),
  ...require("./ContainerClient.js"),
  ...require("./BlobLeaseClient.js"),
  ...require("./sas/AccountSASPermissions.js"),
  ...require("./sas/AccountSASResourceTypes.js"),
  ...require("./sas/AccountSASServices.js"),
  ...require("./BlobBatch.js"),
  ...require("./BlobBatchClient.js"),
  ...require("./sas/BlobSASPermissions.js"),
  ...require("./sas/ContainerSASPermissions.js"),
  ...require("./sas/SASQueryParameters.js"),
  ...require("./generatedModels.js")
});
