// @azure/playwright/dist/commonjs/core/playwrightService.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConnectOptions = exports.createAzurePlaywrightConfig = void 0;
const tslib_1 = require("tslib");
const constants_js_1 = require("../common/constants.js");
const customerConfig_js_1 = tslib_1.__importDefault(require("../common/customerConfig.js"));
const playwrightServiceConfig_js_1 = require("../common/playwrightServiceConfig.js");
const playwrightServiceEntra_js_1 = tslib_1.__importDefault(require("./playwrightServiceEntra.js"));
const utils_js_1 = require("../utils/utils.js");
const messages_js_1 = require("../common/messages.js");
const playwrightServiceUtils_1 = require("./playwrightServiceUtils-cjs.js");
const performOneTimeOperation = (options) => {
    const oneTimeOperationFlag = process.env[constants_js_1.InternalEnvironmentVariables.ONE_TIME_OPERATION_FLAG] === "true";
    if (oneTimeOperationFlag)
        return;
    process.env[constants_js_1.InternalEnvironmentVariables.ONE_TIME_OPERATION_FLAG] = "true";
    if (options?.serviceAuthType === constants_js_1.ServiceAuth.ACCESS_TOKEN) {
        (0, utils_js_1.warnIfAccessTokenCloseToExpiry)();
    }
};
/**
 * @public
 *
 * Generate playwright configuration integrated with Azure Playwright.
 *
 * @param baseConfig - base playwright configuration
 * @param options - additional options for the service
 * @returns PlaywrightConfig
 *
 * @example
 * ```
 * import { defineConfig } from "playwright/test";
 * import { createAzurePlaywrightConfig } from "@azure/playwright";
 * import playwrightConfig from "./playwright.config";
 *
 * export default defineConfig(playwrightConfig, createAzurePlaywrightConfig(playwrightConfig));
 * ```
 *
 * @example
 * ```
 * import { defineConfig } from "playwright/test";
 * import { createAzurePlaywrightConfig, ServiceOS, ServiceAuth } from "@azure/playwright";
 * import playwrightConfig from "./playwright.config";
 * import { DefaultAzureCredential } from '@azure/identity';
 *
 * export default defineConfig(playwrightConfig, createAzurePlaywrightConfig(playwrightConfig, {
 *  credential: new DefaultAzureCredential(),
 *  serviceAuthType: ServiceAuth.ENTRA_ID,
 *  os: ServiceOS.WINDOWS
 * }));
 * ```
 */
const createAzurePlaywrightConfig = (baseConfig, options) => {
    (0, utils_js_1.validatePlaywrightVersion)();
    (0, utils_js_1.validateServiceUrl)();
    // Set environment variable to indicate user is using service config
    process.env[constants_js_1.InternalEnvironmentVariables.USING_SERVICE_CONFIG] = "true";
    const playwrightVersionInfo = (0, utils_js_1.getVersionInfo)((0, utils_js_1.getPlaywrightVersion)());
    const isMultipleGlobalFileSupported = playwrightVersionInfo.major >= 1 && playwrightVersionInfo.minor >= 49;
    if (options?.credential) {
        playwrightServiceEntra_js_1.default.entraIdAccessToken = options.credential;
    }
    // if global setup/teardown is string -
    // 1. if multiple global file is supported, convert it to array
    // 2. wrap playwright-service global setup/teardown with customer provided global setup/teardown
    // if global setup/teardown is array -
    // 1. if multiple global file is not supported, throw error
    // 2. append playwright-service global setup/teardown with customer provided global setup/teardown
    if (baseConfig && baseConfig.globalSetup) {
        if (typeof baseConfig.globalSetup === "string") {
            if (isMultipleGlobalFileSupported) {
                customerConfig_js_1.default.globalSetup = [baseConfig.globalSetup];
            }
            else {
                customerConfig_js_1.default.globalSetup = baseConfig.globalSetup;
            }
        }
        else {
            if (!isMultipleGlobalFileSupported) {
                throw new Error(messages_js_1.ServiceErrorMessageConstants.MULTIPLE_SETUP_FILE_PLAYWRIGHT_VERSION_ERROR.message);
            }
            customerConfig_js_1.default.globalSetup = baseConfig.globalSetup;
        }
    }
    if (baseConfig && baseConfig.globalTeardown) {
        if (typeof baseConfig.globalTeardown === "string") {
            if (isMultipleGlobalFileSupported) {
                customerConfig_js_1.default.globalTeardown = [baseConfig.globalTeardown];
            }
            else {
                customerConfig_js_1.default.globalTeardown = baseConfig.globalTeardown;
            }
        }
        else {
            if (!isMultipleGlobalFileSupported) {
                throw new Error(messages_js_1.ServiceErrorMessageConstants.MULTIPLE_SETUP_FILE_PLAYWRIGHT_VERSION_ERROR.message);
            }
            customerConfig_js_1.default.globalTeardown = baseConfig.globalTeardown;
        }
    }
    const playwrightServiceConfig = playwrightServiceConfig_js_1.PlaywrightServiceConfig.instance;
    playwrightServiceConfig.setOptions(options);
    playwrightServiceConfig.serviceAuthType =
        options?.serviceAuthType || constants_js_1.DefaultConnectOptionsConstants.DEFAULT_SERVICE_AUTH_TYPE;
    const globalFunctions = {};
    if (options?.serviceAuthType === constants_js_1.ServiceAuth.ACCESS_TOKEN) {
        // mpt PAT requested and set by the customer, no need to setup entra lifecycle handlers
        (0, utils_js_1.validateMptPAT)(utils_js_1.exitWithFailureMessage);
    }
    // If multiple global file is supported, append playwright-service global setup/teardown with customer provided global setup/teardown
    if (isMultipleGlobalFileSupported) {
        globalFunctions.globalSetup = [];
        globalFunctions.globalTeardown = [];
        if (customerConfig_js_1.default.globalSetup) {
            globalFunctions.globalSetup.push(...customerConfig_js_1.default.globalSetup);
        }
        if (customerConfig_js_1.default.globalTeardown) {
            globalFunctions.globalTeardown.push(...customerConfig_js_1.default.globalTeardown);
        }
        globalFunctions.globalSetup.push(playwrightServiceUtils_1.globalPaths.setup);
        globalFunctions.globalTeardown.push(playwrightServiceUtils_1.globalPaths.teardown);
    }
    else {
        // If multiple global file is not supported, wrap playwright-service global setup/teardown with customer provided global setup/teardown
        globalFunctions.globalSetup = playwrightServiceUtils_1.globalPaths.setup;
        globalFunctions.globalTeardown = playwrightServiceUtils_1.globalPaths.teardown;
    }
    performOneTimeOperation(options);
    if (options?.useCloudHostedBrowsers === false) {
        console.log("\nRunning tests using local browsers.");
        return {
            ...globalFunctions,
        };
    }
    if (!process.env[constants_js_1.InternalEnvironmentVariables.MPT_CLOUD_HOSTED_BROWSER_USED]) {
        process.env[constants_js_1.InternalEnvironmentVariables.MPT_CLOUD_HOSTED_BROWSER_USED] = "true";
        console.log("\nRunning tests using Playwright workspaces.");
    }
    return {
        use: {
            connectOptions: {
                wsEndpoint: (0, utils_js_1.getServiceWSEndpoint)(playwrightServiceConfig.runId, playwrightServiceConfig.serviceOs, playwrightServiceConfig.apiVersion, playwrightServiceConfig.sourceType),
                headers: {
                    Authorization: `Bearer ${(0, utils_js_1.getAccessToken)()}`,
                    "x-ms-package-version": `@azure/playwright/${(0, utils_js_1.getPackageVersion)()}`,
                },
                timeout: playwrightServiceConfig.connectTimeout,
                exposeNetwork: playwrightServiceConfig.exposeNetwork,
                slowMo: playwrightServiceConfig.slowMo,
            },
        },
        ...globalFunctions,
    };
};
exports.createAzurePlaywrightConfig = createAzurePlaywrightConfig;
/**
 * @public
 *
 * Get connect options required to connect to Azure Playwright's cloud hosted browsers.
 *
 * @param options - additional options for the service
 * @returns BrowserConnectOptions
 *
 * @example
 * ```
 * import playwright, { test, expect, BrowserType } from "@playwright/test";
 * import { getConnectOptions } from "@azure/playwright";
 *
 * test('has title', async ({ browserName }) => {
 *  const { wsEndpoint, options } = await getConnectOptions();
 *  const browser = await (playwright[browserName] as BrowserType).connect(wsEndpoint, options);
 *  const context = await browser.newContext();
 *  const page = await context.newPage();
 *
 *  await page.goto('https://playwright.dev/');
 *  await expect(page).toHaveTitle(/Playwright/);
 *
 *  await page.close();
 *  await context.close();
 *  await browser.close();
 * });
 * ```
 */
const getConnectOptions = async (options) => {
    const playwrightServiceConfig = playwrightServiceConfig_js_1.PlaywrightServiceConfig.instance;
    playwrightServiceConfig.setOptions(options, true);
    performOneTimeOperation(options);
    playwrightServiceConfig.serviceAuthType =
        options?.serviceAuthType || constants_js_1.DefaultConnectOptionsConstants.DEFAULT_SERVICE_AUTH_TYPE;
    let token;
    if (playwrightServiceConfig.serviceAuthType === constants_js_1.ServiceAuth.ENTRA_ID) {
        if (!options?.credential) {
            throw new Error(messages_js_1.ServiceErrorMessageConstants.NO_CRED_ENTRA_AUTH_ERROR.message);
        }
        playwrightServiceEntra_js_1.default.entraIdAccessToken = options.credential;
        token = await (0, utils_js_1.fetchOrValidateAccessToken)(options.credential);
    }
    else if (playwrightServiceConfig.serviceAuthType === constants_js_1.ServiceAuth.ACCESS_TOKEN) {
        (0, utils_js_1.validateMptPAT)(utils_js_1.throwErrorWithFailureMessage);
        token = (0, utils_js_1.getAccessToken)();
    }
    else {
        throw new Error(messages_js_1.ServiceErrorMessageConstants.INVALID_AUTH_TYPE_ERROR.message);
    }
    if (!token) {
        throw new Error(messages_js_1.ServiceErrorMessageConstants.NO_AUTH_ERROR_PAT_TOKEN.message);
    }
    return {
        wsEndpoint: (0, utils_js_1.getServiceWSEndpoint)(playwrightServiceConfig.runId, playwrightServiceConfig.serviceOs, playwrightServiceConfig.apiVersion, playwrightServiceConfig.sourceType),
        options: {
            headers: {
                Authorization: `Bearer ${token}`,
                "x-ms-package-version": `@azure/playwright/${(0, utils_js_1.getPackageVersion)()}`,
            },
            timeout: playwrightServiceConfig.connectTimeout,
            exposeNetwork: playwrightServiceConfig.exposeNetwork,
            slowMo: playwrightServiceConfig.slowMo,
        },
    };
};
exports.getConnectOptions = getConnectOptions;
