// @azure/playwright/dist/commonjs/reporter/playwrightReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const playwrightReporterStorageManager_js_1 = require("../utils/playwrightReporterStorageManager.js");
const utils_js_1 = require("../utils/utils.js");
const logger_js_1 = require("../common/logger.js");
const playwrightServiceConfig_js_1 = require("../common/playwrightServiceConfig.js");
const constants_js_1 = require("../common/constants.js");
const messages_js_1 = require("../common/messages.js");
const PlaywrightServiceClient_js_1 = require("../utils/PlaywrightServiceClient.js");
const getPlaywrightVersion_js_1 = require("../utils/getPlaywrightVersion.js");
/**
 * Azure Playwright Reporter - Uploads generated Playwright test report folder to Azure Storage.
 */
class PlaywrightReporter {
    config;
    workspaceMetadata = null;
    isReportingEnabled = false;
    /**
     * Called when test run begins. Stores configuration for later use and validates serviceAuthType and
     * retrieves workspace metadata.
     * @param config - Playwright test configuration
     */
    async onBegin(config) {
        logger_js_1.coreLogger.info(`Reporter configuration: ${JSON.stringify(config.reporter, null, 2)}`);
        this.config = config;
        // Check Playwright version for reporting feature compatibility
        try {
            const playwrightVersion = (0, getPlaywrightVersion_js_1.getPlaywrightVersion)();
            const playwrightVersionInfo = (0, utils_js_1.getVersionInfo)(playwrightVersion);
            // Check if version is < 1.57
            const isReportingSupportedVersion = playwrightVersionInfo.major > 1 ||
                (playwrightVersionInfo.major === 1 && playwrightVersionInfo.minor >= 57);
            if (!isReportingSupportedVersion) {
                console.error(messages_js_1.ServiceErrorMessageConstants.PLAYWRIGHT_VERSION_TOO_OLD_FOR_REPORTING.message);
                this.isReportingEnabled = false;
                return;
            }
        }
        catch (error) {
            logger_js_1.coreLogger.error(`Failed to get Playwright version: ${error}`);
            console.error(messages_js_1.ServiceErrorMessageConstants.PLAYWRIGHT_VERSION_TOO_OLD_FOR_REPORTING.message);
            this.isReportingEnabled = false;
            return;
        }
        // Check if using service config
        const usingServiceConfig = process.env[constants_js_1.InternalEnvironmentVariables.USING_SERVICE_CONFIG] === "true";
        logger_js_1.coreLogger.info(`Using service config: ${usingServiceConfig}`);
        if (!usingServiceConfig) {
            console.error(messages_js_1.ServiceErrorMessageConstants.REPORTER_REQUIRES_SERVICE_CONFIG.message);
            this.isReportingEnabled = false;
            return;
        }
        // Check if test run creation was successful
        const testRunCreationSuccess = process.env[constants_js_1.InternalEnvironmentVariables.TEST_RUN_CREATION_SUCCESS] === "true";
        if (!testRunCreationSuccess) {
            console.error(messages_js_1.ServiceErrorMessageConstants.REPORTING_STATUS_FAILED.message);
            console.error(messages_js_1.ServiceErrorMessageConstants.REPORTING_TEST_RUN_FAILED.message);
            this.isReportingEnabled = false;
            return;
        }
        // Check authentication
        const playwrightServiceConfig = playwrightServiceConfig_js_1.PlaywrightServiceConfig.instance;
        logger_js_1.coreLogger.info(`Current authentication type: ${playwrightServiceConfig.serviceAuthType}`);
        const isUsingAccessToken = playwrightServiceConfig.serviceAuthType === constants_js_1.ServiceAuth.ACCESS_TOKEN;
        if (isUsingAccessToken) {
            console.error(messages_js_1.ServiceErrorMessageConstants.REPORTER_REQUIRES_ENTRA_AUTH.message);
            this.isReportingEnabled = false;
            return;
        }
        logger_js_1.coreLogger.info("Authentication validation passed - using ENTRA_ID");
        // Validate HTML reporter configuration
        if (!this.validateHtmlReporterConfiguration(config)) {
            return;
        }
        // Get workspace metadata for later use
        try {
            const playwrightServiceApiClient = new PlaywrightServiceClient_js_1.PlaywrightServiceClient();
            this.workspaceMetadata = await playwrightServiceApiClient.getWorkspaceMetadata();
            logger_js_1.coreLogger.info(`Received workspace details: ${JSON.stringify(this.workspaceMetadata, null, 2)}`);
            if (!this.isReportingAllowed(this.workspaceMetadata)) {
                return;
            }
            this.isReportingEnabled = true;
            console.log(messages_js_1.ServiceErrorMessageConstants.REPORTING_ENABLED.message);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            console.error(`${messages_js_1.ServiceErrorMessageConstants.WORKSPACE_METADATA_FETCH_FAILED.message}Error: ${errorMessage} `);
        }
    }
    /**
     * Called when test run ends. Uploads Playwright test report to Azure Storage if reporting is enabled.
     */
    async onEnd() {
        if (this.isReportingEnabled) {
            console.log(messages_js_1.ServiceErrorMessageConstants.COLLECTING_ARTIFACTS.message);
            const uploadResult = await this.uploadHtmlReport();
            if (uploadResult.success) {
                if (uploadResult.partialSuccess &&
                    uploadResult.failedFileDetails &&
                    uploadResult.failedFileDetails.length > 0) {
                    console.log("Warning: Failed to upload the following files:");
                    uploadResult.failedFileDetails.forEach((fileDetail) => {
                        console.log(`  - ${fileDetail.fileName}, ERROR: ${fileDetail.error}`);
                    });
                    console.log(messages_js_1.ServiceErrorMessageConstants.REPORTING_STATUS_PARTIAL.message);
                }
                else {
                    console.log(messages_js_1.ServiceErrorMessageConstants.REPORTING_STATUS_SUCCESS.message);
                }
                // Display portal URL for both full and partial success
                if (this.workspaceMetadata?.resourceId) {
                    const portalUrl = (0, utils_js_1.getPortalTestRunUrl)(this.workspaceMetadata.resourceId);
                    console.log(messages_js_1.ServiceErrorMessageConstants.TEST_REPORT_VIEW_URL.formatWithUrl(portalUrl));
                }
            }
            else {
                console.error(messages_js_1.ServiceErrorMessageConstants.REPORTING_STATUS_FAILED.message);
                if (uploadResult.errorMessage) {
                    console.error(`Error: ${uploadResult.errorMessage}`);
                }
            }
        }
    }
    async uploadHtmlReport() {
        try {
            const outputFolder = (0, utils_js_1.getHtmlReporterOutputFolder)(this.config);
            logger_js_1.coreLogger.info(`Resolved Playwright test report output folder: ${outputFolder}`);
            const storageManager = new playwrightReporterStorageManager_js_1.PlaywrightReporterStorageManager();
            logger_js_1.coreLogger.info("Starting Playwright test report upload process");
            const uploadResult = await storageManager.uploadPlaywrightHtmlReportAfterTests(outputFolder, this.workspaceMetadata);
            if (uploadResult.success) {
                logger_js_1.coreLogger.info(`Playwright Test report uploaded successfully to Azure Storage.`);
            }
            return uploadResult;
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : "Unknown error";
            logger_js_1.coreLogger.error(`Upload failed: ${errorMessage}`);
            return { success: false, errorMessage };
        }
    }
    validateHtmlReporterConfiguration(config) {
        if (!config.reporter || !Array.isArray(config.reporter)) {
            console.error(messages_js_1.ServiceErrorMessageConstants.HTML_REPORTER_REQUIRED.message);
            return false;
        }
        const htmlReporterIndex = config.reporter.findIndex((reporter) => {
            const reporterName = typeof reporter === "string" ? reporter : reporter?.[0];
            return reporterName === "html";
        });
        const azureReporterIndex = config.reporter.findIndex((reporter) => {
            const reporterName = typeof reporter === "string" ? reporter : reporter?.[0];
            return (typeof reporterName === "string" &&
                reporterName.includes("playwright") &&
                reporterName.includes("reporter"));
        });
        // Validate HTML reporter exists
        if (htmlReporterIndex === -1) {
            console.error(messages_js_1.ServiceErrorMessageConstants.HTML_REPORTER_REQUIRED.message);
            return false;
        }
        // Validate HTML reporter comes before Azure reporter (if Azure reporter exists)
        if (azureReporterIndex !== -1 && htmlReporterIndex > azureReporterIndex) {
            console.error(messages_js_1.ServiceErrorMessageConstants.HTML_REPORTER_REQUIRED.message);
            return false;
        }
        logger_js_1.coreLogger.info("HTML reporter validation passed - HTML reporter is configured and properly ordered");
        return true;
    }
    /**
     * Determines if reporting should be enabled based on workspace metadata.
     * @param workspaceMetadata - The workspace metadata containing reporting and storageUri information
     * @returns true if reporting should be enabled, false otherwise
     */
    isReportingAllowed(workspaceMetadata) {
        if (!workspaceMetadata) {
            console.error(messages_js_1.ServiceErrorMessageConstants.FAILED_TO_GET_WORKSPACE_METADATA.message);
            return false;
        }
        const { reporting, storageUri } = workspaceMetadata;
        // If reporting field is present in metadata, check its value
        if (reporting !== undefined) {
            const normalizedReporting = typeof reporting === "string" ? reporting.toLowerCase() : reporting;
            if (normalizedReporting === "disabled") {
                console.error(messages_js_1.ServiceErrorMessageConstants.WORKSPACE_REPORTING_DISABLED.message);
                logger_js_1.coreLogger.info("Reporting disabled via workspace metadata configuration");
                return false;
            }
            if (normalizedReporting === "enabled" && !storageUri) {
                console.error(messages_js_1.ServiceErrorMessageConstants.WORKSPACE_REPORTING_STORAGE_NOT_LINKED.message);
                logger_js_1.coreLogger.info("Reporting enabled in metadata but storage URI not configured");
                return false;
            }
            if (normalizedReporting !== "enabled") {
                logger_js_1.coreLogger.info(`Unexpected reporting value in workspace metadata: ${reporting}. Falling back to storage URI check.`);
            }
        }
        if (!storageUri) {
            console.error(messages_js_1.ServiceErrorMessageConstants.WORKSPACE_REPORTING_DISABLED.message);
            logger_js_1.coreLogger.info("Storage URI not configured in workspace metadata");
            return false;
        }
        if (!(0, utils_js_1.isValidAzureStorageBlobUri)(storageUri)) {
            console.error(messages_js_1.ServiceErrorMessageConstants.INVALID_STORAGE_URI.message);
            logger_js_1.coreLogger.error("Reporting disabled: storageUri is not a valid Azure Storage endpoint.");
            return false;
        }
        logger_js_1.coreLogger.info("Reporting enabled based on storage URI configuration");
        return true;
    }
}
exports.default = PlaywrightReporter;
